src/site/content/en/blog/mishipay/index.md

```diff
@@ -0,0 +1,347 @@
+    const canvasWidth = (canvasEle.width = videoEle.videoWidth);
+    canvasCtx.drawImage(videoEle, 0, 0, canvasWidth, canvasHeight);
+    // Transfer the `canvasEle` to the decoder for barcode detection.
+    const result = await decodeBarcode(canvasEle);
+  } else {
+    console.log('Video feed not available yet');
+  }
+}
+```
+
+For advanced use cases, this layer also performs some pre-processing tasks such as cropping,
+rotation, and converting to grayscale. These tasks can be CPU-intensive and result in the
+application being unresponsive given barcode scanning is a long-running operation. With the help of
+the [OffscreenCanvas](https://developer.mozilla.org/docs/Web/API/OffscreenCanvas) API, we can
+offload the CPU-intensive task to a web worker. On devices that support hardware graphics
+acceleration, leveraging WebGL API and its
+[`WebGL2RenderingContext`](https://developer.mozilla.org/docs/Web/API/WebGL2RenderingContext) can be
+a strategy for some optimization gains on the CPU-intensive pre-processing tasks.
+
+### Decoder layer
+
+The final layer is the decoder layer that is responsible for decoding the barcodes from the frames
+captured by the processing layer. Thanks to the
+[Shape Detection API](https://developer.mozilla.org/docs/Web/API/Barcode_Detection_API),
+starting from Chromium&nbsp;83, the browser itself supports the functionality of decoding a barcode
+from an `ImageBitmapSource`, which can be an `img` element, an SVG `image` element, a `video`
+element, a `canvas` element, a `Blob` object, an `ImageData` object, or an `ImageBitmap` object.
+
+{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/GR4od5fHOxys6lrxttPn.png", alt="Diagram showing the three main thread layers: video stream, processing layer, and Shape Detection API.", width="800", height="358" %}
+
+```js
+/**
+ * Barcode Decoder with Shape Detection API
+ * https://web.dev/shape-detection/
+ */
+async function decodeBarcode(canvas) {
+  const formats = [
+    'aztec',
+    'code_128',
+    'code_39',
+    'code_93',
+    'codabar',
+    'data_matrix',
+    'ean_13',
+    'ean_8',
+    'itf',
+    'pdf417',
+    'qr_code',
+    'upc_a',
+    'upc_e',
+  ];
+  const barcodeDetector = new window.BarcodeDetector({
+    formats,
+  });
+  try {
+    const barcodes = await barcodeDetector.detect(canvas);
+    console.log(barcodes);
+    return barcodes.length > 0 ? barcodes[0]['rawValue'] : undefined;
+  } catch (e) {
+    throw e;
+  }
+}
+```
+
+For devices that don't support the Shape Detection API yet, we would need a fallback solution to
+decode the barcodes. The Shape Detection API exposes a
+[`getSupportedFormats()`](https://developer.mozilla.org/docs/Web/API/BarcodeDetector/getSupportedFormats)
+method that can also be used as a deciding factor by the decoding layer to decide between the Shape
+Detection API and the fallback solution.
+
+```js
+// Check supported barcode formats.
+BarcodeDetector.getSupportedFormats().then((supportedFormats) => {
```

Comment:
```suggestion
BarcodeDetector.getSupportedFormats()
.then((supportedFormats) => {
```

Reply:
We typically wrap the `then()` clause (if we're not using `await`. Aside from consistency, this avoids horizontal scrolling on small screens.

Reply:
Sorry, this was prettier. Overruled it manually.